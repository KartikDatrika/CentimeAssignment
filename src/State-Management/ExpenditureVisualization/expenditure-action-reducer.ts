import { ADD_FLOW, DELETE_FLOW, EDIT_FLOW, GET_EXPENDITURE_DATA, IExpenditureAction } from "./expenditure-action-types";

const initialData = {
    data: [] as any
}

export const EG_INCOME = "eg: Income";
export const EG_EXPENDITURE = "eg: Expenditure";

export default function expenditureReducer(state = initialData, action: IExpenditureAction) {
    
    let newData = action.payload && [...action.payload.data];
    
    switch (action.type) {
        case GET_EXPENDITURE_DATA:
            break;
        case ADD_FLOW:
            newData.splice(1, 0, [EG_INCOME, EG_EXPENDITURE, 1]);
            break;
        case EDIT_FLOW:
            newData[action.payload.x][action.payload.y] = action.payload.value;
            break;
        case DELETE_FLOW:
            newData.splice(action.payload.x, 1);
            break;
        default:
            return state;
    }
    
    return {
        data: newData
    }
}