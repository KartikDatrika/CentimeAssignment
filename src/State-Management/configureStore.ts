import { applyMiddleware, compose, createStore, combineReducers } from 'redux'
import thunkMiddleware from 'redux-thunk'
import expenditureReducer from './ExpenditureVisualization/expenditure-action-reducer'

export interface IAppState{
    expenditureData: {data:any[]}
}

const rootReducer = combineReducers({
    expenditureData: expenditureReducer
})

export default function configureStore() {
    const middlewares = [thunkMiddleware]
    const middlewareEnhancer = applyMiddleware(...middlewares)
    const composedEnhancers = compose(middlewareEnhancer)
    const store = createStore(rootReducer, composedEnhancers);
    return store
}
