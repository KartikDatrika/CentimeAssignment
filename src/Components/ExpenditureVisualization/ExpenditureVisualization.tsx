import { useEffect } from 'react'
import { ReactGoogleChartProps } from 'react-google-charts/dist/types';
import { useDispatch, useSelector } from 'react-redux';
import { IAppState } from '../../State-Management/configureStore';
import { fetchDataFromServer } from '../../State-Management/ExpenditureVisualization/expenditure-action-creators';
import GoogleChart from '../Charts/GoogleChart';
import LanguageSwitcher from '../Utility/LanguageSwitcher/LanguageSwitcher';
import EditableTable from '../Utility/EditableTable/EditableTable';

export default function ExpenditureVisualization() {

    const data = useSelector((state: IAppState) => state.expenditureData.data);
    const dispatch = useDispatch();

    useEffect(() => {
        fetchDataFromServer(dispatch, 'http://localhost:3001/flows');
    }, [])

    let chartProps: ReactGoogleChartProps = {
        chartType: "Sankey",
        height: 500,
        data
    }

    const chart = data.length > 0 ? <GoogleChart {...chartProps} /> : null;

    return (
        <div className="container">
            <LanguageSwitcher/>
            <EditableTable translateColumnNames={true}/>
            {chart}
        </div>
    )
}
