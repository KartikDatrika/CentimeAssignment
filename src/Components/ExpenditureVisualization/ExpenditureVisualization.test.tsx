import React from 'react';
import { fireEvent, render, waitFor } from '@testing-library/react';
import ExpenditureVisualization from './ExpenditureVisualization';
import { Provider } from 'react-redux';
import configureStore from '../../State-Management/configureStore';

describe('ExpenditureVisualization Test Suite', () => {
  let mockStore = {} as any;
  let response: Response = {
    json: [
      [
        "InFlow",
        "OutFlow",
        "Amount"
      ],
      [
        "Stocks",
        "Savings",
        25
      ],
      [
        "Gold",
        "Savings",
        10
      ]]
  } as any;
  beforeAll(() => {
    mockStore = configureStore();
    jest.spyOn(window, 'fetch');
  })

  it('ExpenditureVisualization is rendering with editable table having 3 rows !', async () => {
    window.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => (response.json),
    })
    const { container, getAllByRole } = render(<Provider store={mockStore}><ExpenditureVisualization /></Provider>);
    await waitFor(() => expect(window.fetch).toHaveBeenCalledTimes(1));
    expect(container).toMatchSnapshot();
    const b = getAllByRole("button");
    expect(b).toHaveLength(3);
  });

  it('on Add row gets added , on del row gets deleted and can edit the cell !', async () => {
    window.fetch.mockResolvedValueOnce({
      ok: true,
      json: async () => (response.json),
    })
    const { getByDisplayValue, getAllByRole } = render(<Provider store={mockStore}><ExpenditureVisualization /></Provider>);
    await waitFor(() => expect(window.fetch).toHaveBeenCalledTimes(1));
    const rows = getAllByRole("button");
    const add = rows[0];
    const del = rows[1];
    add.click()
    expect(getAllByRole("button")).toHaveLength(4);
    del.click()
    expect(getAllByRole("button")).toHaveLength(3);
    const inputCell = getByDisplayValue("Stocks");
    expect(fireEvent.change(inputCell, { target: { value: "stocks" } }));
  });

  it('on api fail app renders without crash', () => {
    const { container} = render(<Provider store={mockStore}><ExpenditureVisualization /></Provider>);
    expect(container).toMatchSnapshot();
  })
})
