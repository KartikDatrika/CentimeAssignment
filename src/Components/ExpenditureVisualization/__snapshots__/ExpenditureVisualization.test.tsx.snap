// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenditureVisualization Test Suite ExpenditureVisualization is rendering with editable table having 3 rows ! 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="mb-5"
    >
      <label
        class="columnName"
      >
        Translate to
      </label>
      <select
        class="form-control"
        data-testid="select"
      >
        <option
          data-testid="select-option"
          value="en"
        >
          English
        </option>
        <option
          data-testid="select-option"
          value="af"
        >
          Afrikaans
        </option>
        <option
          data-testid="select-option"
          value="bg"
        >
          Bulgarian
        </option>
        <option
          data-testid="select-option"
          value="hr"
        >
          Croatian
        </option>
        <option
          data-testid="select-option"
          value="cs"
        >
          Czech
        </option>
        <option
          data-testid="select-option"
          value="da"
        >
          Danish
        </option>
        <option
          data-testid="select-option"
          value="nl"
        >
          Dutch
        </option>
        <option
          data-testid="select-option"
          value="fr"
        >
          French
        </option>
        <option
          data-testid="select-option"
          value="de"
        >
          German
        </option>
        <option
          data-testid="select-option"
          value="el"
        >
          Greek
        </option>
        <option
          data-testid="select-option"
          value="hi"
        >
          Hindi
        </option>
        <option
          data-testid="select-option"
          value="ja"
        >
          Japanese
        </option>
        <option
          data-testid="select-option"
          value="ko"
        >
          Korean
        </option>
        <option
          data-testid="select-option"
          value="ru"
        >
          Russian
        </option>
        <option
          data-testid="select-option"
          value="sr"
        >
          Serbian
        </option>
        <option
          data-testid="select-option"
          value="sk"
        >
          Slovak
        </option>
        <option
          data-testid="select-option"
          value="es"
        >
          Spanish
        </option>
        <option
          data-testid="select-option"
          value="sv"
        >
          Swedish
        </option>
        <option
          data-testid="select-option"
          value="th"
        >
          Thai
        </option>
        <option
          data-testid="select-option"
          value="tr"
        >
          Turkish
        </option>
        <option
          data-testid="select-option"
          value="vi"
        >
          Vietnamese
        </option>
      </select>
    </div>
    <div
      class="table-container"
    >
      <table
        class="table table-borderless"
      >
        <thead>
          <tr>
            <td
              class="columnName"
            >
               
              InFlow
               
            </td>
            <td
              class="columnName"
            >
               
              OutFlow
               
            </td>
            <td
              class="columnName"
            >
               
              Amount
               
            </td>
            <td>
              <button
                class="btn btn-success"
                type="button"
              >
                ADD
              </button>
            </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Stocks"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Savings"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="25"
              />
            </td>
            <td>
              <button
                class="btn btn-danger"
                type="button"
              >
                DEL
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Gold"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Savings"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="10"
              />
            </td>
            <td>
              <button
                class="btn btn-danger"
                type="button"
              >
                DEL
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="mt-5 mb-5"
    />
  </div>
</div>
`;

exports[`ExpenditureVisualization Test Suite on api fail app renders without crash 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="mb-5"
    >
      <label
        class="columnName"
      >
        Translate to
      </label>
      <select
        class="form-control"
        data-testid="select"
      >
        <option
          data-testid="select-option"
          value="en"
        >
          English
        </option>
        <option
          data-testid="select-option"
          value="af"
        >
          Afrikaans
        </option>
        <option
          data-testid="select-option"
          value="bg"
        >
          Bulgarian
        </option>
        <option
          data-testid="select-option"
          value="hr"
        >
          Croatian
        </option>
        <option
          data-testid="select-option"
          value="cs"
        >
          Czech
        </option>
        <option
          data-testid="select-option"
          value="da"
        >
          Danish
        </option>
        <option
          data-testid="select-option"
          value="nl"
        >
          Dutch
        </option>
        <option
          data-testid="select-option"
          value="fr"
        >
          French
        </option>
        <option
          data-testid="select-option"
          value="de"
        >
          German
        </option>
        <option
          data-testid="select-option"
          value="el"
        >
          Greek
        </option>
        <option
          data-testid="select-option"
          value="hi"
        >
          Hindi
        </option>
        <option
          data-testid="select-option"
          value="ja"
        >
          Japanese
        </option>
        <option
          data-testid="select-option"
          value="ko"
        >
          Korean
        </option>
        <option
          data-testid="select-option"
          value="ru"
        >
          Russian
        </option>
        <option
          data-testid="select-option"
          value="sr"
        >
          Serbian
        </option>
        <option
          data-testid="select-option"
          value="sk"
        >
          Slovak
        </option>
        <option
          data-testid="select-option"
          value="es"
        >
          Spanish
        </option>
        <option
          data-testid="select-option"
          value="sv"
        >
          Swedish
        </option>
        <option
          data-testid="select-option"
          value="th"
        >
          Thai
        </option>
        <option
          data-testid="select-option"
          value="tr"
        >
          Turkish
        </option>
        <option
          data-testid="select-option"
          value="vi"
        >
          Vietnamese
        </option>
      </select>
    </div>
    <div
      class="table-container"
    >
      <table
        class="table table-borderless"
      >
        <thead>
          <tr>
            <td
              class="columnName"
            >
               
              InFlow
               
            </td>
            <td
              class="columnName"
            >
               
              OutFlow
               
            </td>
            <td
              class="columnName"
            >
               
              Amount
               
            </td>
            <td>
              <button
                class="btn btn-success"
                type="button"
              >
                ADD
              </button>
            </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="stocks"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Savings"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="25"
              />
            </td>
            <td>
              <button
                class="btn btn-danger"
                type="button"
              >
                DEL
              </button>
            </td>
          </tr>
          <tr>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Gold"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="Savings"
              />
            </td>
            <td>
              <input
                class="form-control"
                placeholder=""
                value="10"
              />
            </td>
            <td>
              <button
                class="btn btn-danger"
                type="button"
              >
                DEL
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="mt-5 mb-5"
    />
  </div>
</div>
`;
